### ISSUE TITLE
Support variables passed by reference in function calls. 

### ISSUE THREAD

**Issue by parth-07** [vgvassilev/clad#354]
Currently, we do not correctly differentiate variables passed by reference in function calls. We do not provide any warning/error either for this. 

Minimum reproducible example:
```c++
void square_inplace(double& i) {
  i = i*i;
}

double fn(double i, double j) {
  square_inplace(i);
  return i;
}

auto d_fn = clad::gradient(fn);
double d_i, d_j;
d_i = d_j = 0;
d_fn.execute(3, 5, &d_i, &d_j);
```
**Result:** d_i: 1, d_j: 0
**Correct Result:** d_i: 6, d_j: 0

We can relatively add support for variables passed by reference in both the forward and the reverse differentiation mode. For forward mode, this support will be added as part of PR #345, and for reverse mode, we can simply pass derived parameters by reference whenever the corresponding original parameter is being passed by reference, just this should add this support.

**Comment by grimmmyshini** [comment]
Please also consider syncing this with #281.

**Comment by parth-07** [comment]
Fixed with commit c35bcff7457801fcfdba7f6111ced8290d85a24f

**Comment by parth-07** [comment]
Fixed with commit c35bcff7457801fcfdba7f6111ced8290d85a24f
